src/Mod/Draft/draftutils/grid_observer.py

```diff
@@ -0,0 +1,133 @@
+                            "draft_snap_widget", "Toggle grid", False, False
+                        )
+                return
+            _update_gridgui()
+        except Exception:
+            pass
+
+    def _view_observer_callback(sub_win):
+        if not hasattr(FreeCADGui, "draftToolBar"):
+            return
+        tray = FreeCADGui.draftToolBar.tray
+        if tray is None:
+            return
+        if FreeCADGui.draftToolBar.tray.isVisible() is False:
+            return
+        ToDo.delay(_update_gui, None)
+
+    _view_observer_active = False
+
+    def _view_observer_start():
+        mw = FreeCADGui.getMainWindow()
+        mdi = mw.findChild(QtWidgets.QMdiArea)
+        global _view_observer_active
+        if not _view_observer_active:
+            mdi.subWindowActivated.connect(_view_observer_callback)
+            _view_observer_active = True
+            _view_observer_callback(mdi.activeSubWindow())  # Trigger initial update.
+
+    def _view_observer_stop():
+        mw = FreeCADGui.getMainWindow()
+        mdi = mw.findChild(QtWidgets.QMdiArea)
+        global _view_observer_active
+        if _view_observer_active:
+            mdi.subWindowActivated.disconnect(_view_observer_callback)
+            _view_observer_active = False
+
+    def _update_gridgui():
+        if (
+            FreeCAD.GuiUp
+            and hasattr(FreeCADGui, "draftToolBar")
+            and gui_utils.get_3d_view() is not None
+        ):
+            mw = FreeCADGui.getMainWindow()
+            if mw.findChild(QtWidgets.QToolBar, "Draft snap"):
+                if FreeCADGui.Snapper.grid.Visible:
+                    tbSnap = mw.findChild(QtWidgets.QToolBar, "Draft snap")
+                    btnGrid = _find_grid_toolbutton(
+                        "Draft snap", "Toggle grid", True, True
+                    )
+                    if mw.findChild(QtWidgets.QToolBar, "draft_snap_widget"):
+                        tbSnapWidget = mw.findChild(
+                            QtWidgets.QToolBar, "draft_snap_widget"
+                        )
+                        btnGridSW = _find_grid_toolbutton(
+                            "draft_snap_widget", "Toggle grid", True, True
+                        )
+                else:
+                    tbSnap = mw.findChild(QtWidgets.QToolBar, "Draft snap")
+                    btnGrid = _find_grid_toolbutton(
+                        "Draft snap", "Toggle grid", True, False
+                    )
+                    if mw.findChild(QtWidgets.QToolBar, "draft_snap_widget"):
+                        tbSnapWidget = mw.findChild(
+                            QtWidgets.QToolBar, "draft_snap_widget"
+                        )
+                        btnGridSW = _find_grid_toolbutton(
+                            "draft_snap_widget", "Toggle grid", True, False
+                        )
+
+    def _find_grid_toolbutton(tbObj, tbButton, tbButtonEnabled, tbButtonChecked):
```

Comment:
The function `_find_grid_toolbutton` name is confusing. The function does not return a button. The fact that `tbObj` is a string is also a bit strange IMO, `toolbar_name` and `button_name` would be clearer.

Reply:
Fixed with last commit.